packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,293 @@
+import { DataFrame, Field, FieldMatcher, FieldType, Vector } from '../../types';
+import { ArrayVector } from '../../vector';
+import { fieldMatchers } from '../matchers';
+import { FieldMatcherID } from '../matchers/ids';
+import { getTimeField, sortDataFrame } from '../../dataframe';
+import { getFieldDisplayName } from '../../field';
+
+export function pickBestJoinField(data: DataFrame[]): FieldMatcher {
+  const { timeField } = getTimeField(data[0]);
+  if (timeField) {
+    return fieldMatchers.get(FieldMatcherID.firstTimeField).get({});
+  }
+  let common: string[] = [];
+  for (const f of data[0].fields) {
+    if (f.type === FieldType.number) {
+      common.push(f.name);
+    }
+  }
+
+  for (let i = 1; i < data.length; i++) {
+    const names: string[] = [];
+    for (const f of data[0].fields) {
+      if (f.type === FieldType.number) {
+        names.push(f.name);
+      }
+    }
+    common = common.filter((v) => !names.includes(v));
+  }
+
+  return fieldMatchers.get(FieldMatcherID.byName).get(common[0]);
+}
+
+/**
+ * @alpha
+ */
+export interface JoinOptions {
+  /**
+   * The input fields
+   */
+  frames: DataFrame[];
+
+  /**
+   * The field to join -- frames that do not have this field will be droppped
+   */
+  joinBy?: FieldMatcher;
+
+  /**
+   * Optionally filter the non-join fields
+   */
+  keep?: FieldMatcher;
+
+  /**
+   * Make sure the results are always sorted with the `joinBy` field
+   */
+  enforceSort?: boolean;
+
+  /**
+   * @internal -- used when we need to keep a reference to the original frame/field index
+   */
+  keepOriginIndices?: boolean;
+}
+
+/**
+ * This will return a single frame joined by the first matching field.  When a join field is not specified,
+ * the default will use the first time field
+ */
+export function outerJoinDataFrames(options: JoinOptions): DataFrame | undefined {
+  if (!options?.frames?.length) {
```

Comment:
neither options nor frames is optional here, so drop the `?` ?

Reply:
I droped the first, but will keep the second -- although the types require it to be set, the runtime environment does not.  Since this function gets called often on raw data it seems like a cheap way to avoid exploding

Reply:
so then it's missing on `frames` here?: https://github.com/grafana/grafana/pull/30407#pullrequestreview-578023461

Reply:
no -- since we already checked it is valid, it does not need to be checked again.

Interestingly, if `frames` was defined as optional, typescript would know it is  not optional after this point